tests/src/test/scala/com/nvidia/spark/rapids/HashAggregatesSuite.scala

```diff
@@ -637,6 +638,7 @@ class HashAggregatesSuite extends SparkQueryCompareTestSuite {
   FLOAT_TEST_testSparkResultsAreEqual(
       "doubles basic aggregates group by doubles",
       doubleCsvDf,
+      maxFloatDiff = 0.0001,
```

Comment:
Aggregation ordering isn't deterministic, so I understand why this change could be necessary, but I'm not sure why we weren't seeing this before.  Are we sure the exact same values from the CSV file are being loaded as before this PR and this change to the test isn't masking another problem?

Reply:
For some reason, this seems to happen pretty reliably with Spark 3.1.1 (I had previously only tested locally with 3.0.1 and 3.2.0). Here is a subset of the results showing the difference between CPU and GPU.

## CPU

```
WrappedArray(-100.0, 456.0, 23.0, 129.0, -600.0, -106.0, 106.0, -94.0, -100.0, 1), 
WrappedArray(100.0, 456.0, 223.0, 124.0, 100.0, 99.0, -99.0, 101.0, 100.0, 1), 
WrappedArray(200.0, 456.0, 323.0, 125.0, 400.0, 198.0, -198.0, 202.0, 200.0, 1), 
WrappedArray(300.0, 456.0, 423.0, 126.0, 900.0, 297.0, -297.0, 303.0, 300.0, 1), 
WrappedArray(400.0, 456.0, 523.0, 127.0, 1600.0, 396.0, -396.0, 404.0, 400.0, 1), 
```

## GPU

```
WrappedArray(-100.0, 456.0, 23.0, 129.0, -599.9999999999999, -106.0, 106.0, -94.0, -100.0, 1), 
WrappedArray(100.0, 456.0, 223.0, 124.0, 99.99999999999999, 99.0, -99.0, 101.0, 100.0, 1), 
WrappedArray(200.0, 456.0, 323.0, 125.0, 399.99999999999994, 198.0, -198.0, 202.0, 200.0, 1), 
WrappedArray(300.0, 456.0, 423.0, 126.0, 899.9999999999999, 297.0, -297.0, 303.0, 300.0, 1), 
WrappedArray(400.0, 456.0, 523.0, 127.0, 1599.9999999999998, 396.0, -396.0, 404.0, 400.0, 1),
```

Reply:
No, this is definitely related to this PR. The affected column is `max(col("doubles") * col("more_doubles"))` ... I will dig into this more and see if I can determine why this is only causing issues with Spark 3.1.1

Reply:
I think the first thing to check is what I asked earlier:
> Are we sure the exact same values from the CSV file are being loaded as before this PR

Reply:
> Are we sure the exact same values from the CSV file are being loaded as before this PR

We know that they will not be the same.  We know that CUDF is not parsing them the same as Spark is, and we also taking steps to change how we parse the numbers when casting to try to make it closer to Spark. Until we have code to parse it exactly the same as Spark, I don't know how much we care except that it is close enough.

We had a huge discussion earlier about what close enough means for floating point.  In the pytest code we typically go with the pytest default relative tolerance of `1e-6`. Or we can be off by one part per million.

I am much more interested to understand if both we and Spark are producing the same results each time? And if Spark is changing (because it looks like 3.1.1 is different from the others) why is it changing?

Reply:
So it turns out that this is not specific to 3.1.1 (I saw the same failure with 3.0.1) and is an intermittent issue, presumably dependent on the order that accumulations are happening during the aggregate. I added debug logging and confirmed that the values read from CSV and the resulting CAST values are the same for 3.0.1 and 3.1.1.

Reply:
This PR was originally motivated for JSON support (especially for handling of out-of-range numbers and NaN/Inf parsing) but also impacts CSV since this is common code. Technically, this is a regression in this case because of the small difference between native cuDF parsing vs casting from string, but there are other cases where these changes make CSV support more compatible with Spark. We could consider putting this code being a configuration option, or just having this change apply to JSON and not CSV if this is a concern.

Reply:
Yes, ultimately this will be much better for some values and slightly worse for others.  It is quite odd that cudf is parsing CSV floating point values directly differently than casting the strings as floating point.

Once the tolerances are lowered to match pytest, I'm good to go on this.